### PR TITLE
module: moving TFM CMakeLists.txt into Zephyr repo

### PR DESCRIPTION
**Pull request by tejlmand**
This commit moves TFM CMakeLists.txt into Zephyr and relocates the
Kconfig.tfm file under the modules/tfm folder.

Updates the manifest to point to related TFM changes.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

--------
~This PR is depending on https://github.com/zephyrproject-rtos/zephyr/pull/30904~
~This PR is depending on https://github.com/zephyrproject-rtos/zephyr/pull/31422~

### CODE REVIEW

File: modules/modules.cmake [hunk 1]
```diff
@@ -0,0 +1,17 @@
+file(GLOB cmake_modules "${CMAKE_CURRENT_LIST_DIR}/*/CMakeLists.txt")
```
Comment (hakonfam) [inline modules/modules.cmake:1]: Should there be a license tag here?

File: modules/Kconfig [hunk 2]
```diff
@@ -30,3 +34,23 @@ source "modules/Kconfig.tinycbor"
...
+
+comment "Unavailable modules, please install those via the project manifest."
+
+# List of Zephyr modules text when module is not available, please use the
+# following syntax:
+# ---------------------------------------------------
+# comment "<module_name> module not available."
+#	depends on !ZEPHYR_<MODULE_NAME_UPPER>_MODULE
+#
+# Remember to add the following code inside the `<module>/Kconfig.<module> file:
```
Comment (hakonfam) [inline modules/Kconfig:45]: `<module_root>/<module>/Kconfig.<module>` or have I misunderstood?

File: modules/Kconfig [hunk 3]
```diff
@@ -30,3 +34,23 @@ source "modules/Kconfig.tinycbor"
 source "modules/Kconfig.tinycrypt"
 source "modules/Kconfig.vega"
 source "modules/Kconfig.xtensa"
+
+comment "Unavailable modules, please install those via the project manifest."
+
+# List of Zephyr modules text when module is not available, please use the
+# following syntax:
+# ---------------------------------------------------
+# comment "<module_name> module not available."
```
Comment (hakonfam) [inline modules/Kconfig:42]: Could a template be used for this?

File: modules/modules.cmake [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+foreach(module ${cmake_modules})
+  get_filename_component(module_dir  ${module} DIRECTORY)
+  get_filename_component(module_name ${module_dir} NAME)
+  string(TOUPPER ${module_name} MODULE_NAME_UPPER)
+  set(ZEPHYR_${MODULE_NAME_UPPER}_CMAKE_DIR ${module_dir})
+endforeach()
+
+file(GLOB kconfig_modules "${CMAKE_CURRENT_LIST_DIR}/*/Kconfig.*")
+
+foreach(module ${cmake_modules})
```
Comment (hakonfam) [inline modules/modules.cmake:12]: `foreach(module ${kconfig_modules})` ?

File: cmake/zephyr_module.cmake [hunk 5]
```diff
@@ -69,6 +81,7 @@ if(WEST OR ZEPHYR_MODULES)
       # Match "<name>":"<path>" for each line of file, each corresponding to
       # one module. The use of quotes is required due to CMake not supporting
       # lazy regexes (it supports greedy only).
+      string(CONFIGURE ${module} module)
```
Comment (hakonfam) [inline cmake/zephyr_module.cmake:84]: Why is this needed now?

File: cmake/zephyr_module.cmake [hunk 6]
```diff
@@ -48,6 +48,8 @@ if(WEST OR ZEPHYR_MODULES)
       message(FATAL_ERROR "${zephyr_module_error_text}")
   endif()
 
+  set(MODULE_ROOT ${ZEPHYR_BASE}/modules)
```
Comment (hakonfam) [inline cmake/zephyr_module.cmake:51]: Should the comment in the top of this file be updated to include a description of the `MODULE_ROOD` functionality?

File: modules/modules.cmake (hunk 1 above)
Comment (tejlmand) [reply modules/modules.cmake:1]: of course.

Added.

File: modules/Kconfig (hunk 2 above)
Comment (tejlmand) [reply modules/Kconfig:45]: This refers to this folder in which this file is used.

For external modules roots, that module is free to define how to handle this Kconfig stuff.
All modules natively supported in Zephyr uses this config to handle correct
```
comment "<module_name> module not available."
```
presentation, when module is not available.
But custom modules not natively in Zephyr is free not to do this.

File: modules/Kconfig (hunk 3 above)
Comment (tejlmand) [reply modules/Kconfig:42]: Then you would end with three lines instead of two, like:
```
module_name=<module_name>
MODULE_NAME_UPPER=<MODULE_NAME_UPPER>
source 'comment.template'
```
plus the template code file, instead of two lines.

I don't see the benefit in that.

File: modules/modules.cmake (hunk 4 above)
Comment (tejlmand) [reply modules/modules.cmake:12]: of course.

fixed.

File: modules/tfm/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,215 @@
+# Copyright (c) 2019, 2020 Linaro
```
[the remark this answers is not in the document]
Comment (tejlmand) [reply modules/tfm/CMakeLists.txt:1]: it's copied verbatim.

File: cmake/zephyr_module.cmake (hunk 5 above)
Comment (tejlmand) [reply cmake/zephyr_module.cmake:87]: because the `zephyr_module.py` will now place the CMake variable `${ZEPHYR_<MODULE_NAME>_CMAKE_DIR}` in the `zephyr_modules.txt` file when the `cmake-extern: True` is used in module.yml.

Doing it this way, allows a module root handler to set the correct value of  `${ZEPHYR_<MODULE_NAME>_CMAKE_DIR}` to point to the glue code.

When CMake code was always located in the module repo, then the path was known, and hence no need for variables.

File: cmake/zephyr_module.cmake (hunk 6 above)
Comment (tejlmand) [reply cmake/zephyr_module.cmake]: It should.

**Comment by tejlmand** [comment]
Please place review comments regarding the module root feature itself in this PR: #30904

File: cmake/zephyr_module.cmake (hunk 6 above)
Comment (tejlmand) [reply cmake/zephyr_module.cmake]: done.
please make additional comments in #30904

**Review by hakonfam — APPROVED** [review]

**Review by oyvindronningstad — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by carlescufi** [comment]
Buildkite passed with extended timeout:
image

**Review by carlescufi — APPROVED** [review]
